Kube-proxy: use type net.IP for addresses

uablrek opened the pull request:
#### What type of PR is this?

/kind bug
/sig network
/area kube-proxy

#### What this PR does / why we need it:

Addresses for LoadBalancerVIPs, ExternalIPs and LoadBalancerSourceRanges were stored as strings internally. These could be non-canonical which causes problems for proxy-mode=nftables.

To store addresses as strings is generally a bad idea, and `ClusterIP` is already stored as net.IP.

#### Which issue(s) this PR fixes:

Fixes #122611

#### Special notes for your reviewer:

This PR replaces https://github.com/kubernetes/kubernetes/pull/122620, please read it for the background.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

N/A

uablrek committed:
kube-proxy: store LoadBalancerVIPs as net.IP

They were stored as strings which could be non-canonical
and cause problems

uablrek commented:
/cc @danwinship @aojea @aroradaman

uablrek commented:
Shall I try to go all-in, and use [netip.Addr](https://pkg.go.dev/net/netip)?

uablrek commented:
TODO: There are several minor optimizations where String() can be called once and stored. And possibly a major optimization in ipvs by using net.IP in `utilipset.Entry`.

uablrek commented:
/retest

uablrek committed:
kube-proxy: store ExternalIPs as net.IP

They were stored as strings which could be non-canonical
and cause problems

aojea commented:
I remember some PRs of people replacing the net.IP by strings to avoid the conversion performance penalty, but I can't say how much is this penalty.

> and use [netip.Addr](https://pkg.go.dev/net/netip)?

I think we should https://github.com/kubernetes/kubernetes/issues/106161

uablrek commented:
> I remember some PRs of people replacing the net.IP by strings to avoid the conversion performance penalty, but I can't say how much is this penalty.

In this particular case, I would say, not much. The vast majority of services don't have loadBalancerIPs (or externalIPs) I think, and the ClusterIP is already stored as net.IP. And as Dan pointed out, there are a lot of "hidden" conversions, e.g. to check family, that we might get rid of.

uablrek commented:
/retest

uablrek commented:
> and use [netip.Addr](https://pkg.go.dev/net/netip)?

I will finish this PR with net.IP, then make a try to switch net.IP -> netip.Addr, and see where it goes.

aroradaman commented on pkg/proxy/serviceport.go:
```diff
@@ -78,10 +78,10 @@ type BaseServicePortInfo struct {
 	protocol                 v1.Protocol
 	nodePort                 int
-	loadBalancerVIPs         []string
+	loadBalancerVIPs         []net.IP
```
@uablrek we can parse the IPs, stringify again to get the canonical address, and keep the data type as string only?
I think we pass IPs as strings only to the majority of the external APIs with which kube-proxy interacts. 

or maybe an aliased type `IP` (which is just a plain string)  to distinguish it from string everywhere.

uablrek replied:
Please read the discussion in https://github.com/kubernetes/kubernetes/pull/122620#issuecomment-1879730388 and onward

aroradaman replied:
oh, I missed that PR.

uablrek commented:
@danwinship It doesn't look like `ip` in `BaseEndpointInfo` should be a `net.IP`. It is assigned from `endpoint.Addresses[0]` (only).

```go
type BaseEndpointInfo struct {
	// Cache this values to improve performance
	ip   string
// ...
}

danwinship commented:
> Shall I try to go all-in, and use [netip.Addr](https://pkg.go.dev/net/netip)?

I was going to suggest that before, but the problem is that we need "sloppy" parsing (for IPs with leading 0s), and we don't currently have that for `netip.Addr` (and you can't even use `netutils.ParseIPSloppy` and then naively convert to `netip.Addr` because [that doesn't work right](https://github.com/kubernetes/kubernetes/pull/122549/commits/74760445196e444b3bc4bb15e1fc1970c62b9564#diff-8ccdf299272f29f007cdba9f7275f3582de759dfb6d1e3f967ce592c3c67a4dbR30-R44).)

https://github.com/kubernetes/kubernetes/pull/122549 adds a bunch of new helpers to start us off on the path of migrating everything to the netip types... I think we should wait for that before trying to use them here.

danwinship commented on pkg/proxy/servicechangetracker_test.go:
```diff
@@ -86,6 +87,13 @@ func makeServicePortName(ns, name, port string, protocol v1.Protocol) ServicePor
+func makeIPs(ipStr ...string) []net.IP {
+	var ips []net.IP
+	for _, s := range ipStr {
+		ips = append(ips, net.ParseIP(s))
```
verify is going to complain about this; you have to always use `netutils.ParseIPSloppy`.

also, given that every caller of this function only passes a single IP, it might be simpler to just do them all inline? `[]string{"10.1.2.4"}` → `[]net.IP{netutils.ParseIPSloppy("10.1.2.4")}` ?

uablrek replied:
Updated, but I kept the function. The use of a slice indicates that there *may* be more than one IP, and then the function will be handy

danwinship commented on pkg/proxy/servicechangetracker_test.go:
```diff
@@ -580,7 +580,7 @@ func TestServiceToServiceMap(t *testing.T) {
 					svcInfo.protocol != expectedInfo.protocol ||
 					svcInfo.healthCheckNodePort != expectedInfo.healthCheckNodePort ||
-					!sets.New[string](svcInfo.externalIPs...).Equal(sets.New[string](expectedInfo.externalIPs...)) ||
+					!reflect.DeepEqual(svcInfo.externalIPs, expectedInfo.externalIPs) ||
```
In the general case, this is not safe, because there are two ways of representing IPv4 `net.IP`s (the `.To4()` way and the `.To16()` way). So this is not only requiring them to be equal, it's requiring the expected values to have been generated in the same way as the actual values...

Given that we hope to eventually move to `netip.Addr` (which doesn't have this problem), this is probably fine.

uablrek replied:
The problem was that net.IP isn't "comparable", so you can't create a set from them. I know this was a short-cut, but as you say:
> Given that we hope to eventually move to netip.Addr (which doesn't have this problem)

netip.Addr *is* comparable.

uablrek commented:
The LoadBalancerSourceRanges are the largest commit, but feels like the largest improvement both in code quality and performance. Rviewers, please pay extra attention to this one. Among other things, CIDRs are now also cononicalized, like:
```
2002::1234:abcd:ffff:c0a8:101/64 -> 2002:0:0:1234::/64
192.168.200.2/24 -> 192.168.200.0/24
```
To me this feels like a bug-fix, but according to [Hyrum's law](https://www.hyrumslaw.com/), it will trash something somewhere.

aojea commented on cmd/kube-proxy/app/server_linux_test.go:
```diff
@@ -43,6 +43,11 @@ import (
 )
 
+func parseCIDR(cidrStr string) *net.IPNet {
+	_, cidr, _ := netutils.ParseCIDRSloppy(cidrStr)
```
ignoring the error does not sound right, we can pass the t.Testing and fail or do mustParseCIDR(c string) and panic inside

aojea replied:
I prefer the later as is how netip does for tests https://cs.opensource.google/go/go/+/refs/tags/go1.21.5:src/net/netip/netip.go;l=132

uablrek replied:
That is intentional. cidr becomes nil on an error, and that is actually tested in some test-cases

uablrek replied:
Sorry, I confused it with pkg/proxy/util/iptables/traffic_test.go, where it _is_ intentional. Now updated as you suggested.

danwinship commented:
Please keep the commits logically organized in their final form rather than having fixup commits. (So, like, fix the `net.ParseIP`→`netutils.ParseIPSloppy` in the original commits rather than having a separate commit for that, and likewise the linter fixes.)

(I think it's more normal in k8s to put unit test changes along with the corresponding code changes too.)

uablrek commented:
/retest

danwinship commented on pkg/proxy/util/utils_test.go:
```diff
@@ -383,55 +383,55 @@ func TestMapCIDRsByIPFamily(t *testing.T) {
 			wantIPv6:        false,
 			expectCorrect:   nil,
-			expectIncorrect: []string{"fd00:20::1/64"},
+			expectIncorrect: []string{"fd00:20::/64"},
```
Ugh, maybe fix all these (and the ones in `traffic_test.go`) as a separate commit before this one? (The fixes are correct even before the change to `NewDetectLocalByCIDR`; we should not have been testing it using bad CIDRs.)

uablrek replied:
Collected in the first commit in this PR, and I also made a new PR https://github.com/kubernetes/kubernetes/pull/122655, if we want better history tracking. But please lgtm that PR fast if we shall use it.

uablrek committed:
kube-proxy: don't use invalid cidrs in unit test

CIDRs like 192.168.200.3/24 and fd00:20::1/64 replaced with
192.168.200.0/24 and fd00:20::/64

uablrek committed:
kube-proxy: LoadBalancerSourceRanges as *net.IPNet

uablrek commented on pkg/proxy/util/iptables/traffic.go:
```diff
@@ -77,22 +78,21 @@ type detectLocalByCIDR struct {
-	_, parsed, err := netutils.ParseCIDRSloppy(cidr)
-	if err != nil {
-		return nil, err
+func NewDetectLocalByCIDR(cidr *net.IPNet) (LocalTrafficDetector, error) {
```
Good comment! It became much better.

uablrek commented:
/retest

danwinship commented:
/lgtm
/approve

i'll let this race with #122655... :slightly_smiling_face:

uablrek commented:
/retest
